### ISSUE TITLE
fully incremental builds for core platform builds 

### ISSUE THREAD

**Issue by jrhizor** [airbytehq/airbyte#4300]
The following cleanups should fix a lot of our build problems and allow us to move to much faster builds:

* python builds (venv management/incrementality)
   * we could try a central python env using the better supported pygradle. There are some adjacent bug reports about incrementally but it is supported in theory.
   * most of our python slowness/difficulty in making it cacheable and incremental is directly related to the need to have one independent venv for every python connector
   * if these are [split out](https://github.com/airbytehq/airbyte/issues/3828) we may be able to avoid this entirely.
* java builds
   * trivial
* docker builds
   * the bmuschko docker plugin supports incremental builds
* dependencies on docker builds
   * this is handled pretty well by something like the bmuschko docker plugin if we switch to specifying the docker image name and tag in gradle instead of inside the docker file
   * we don't have any complicated cases in the core platform
* licensing / formatting
   * this is a big question mark atm
* seed generation
   * needs to specify inputs and outputs
* json2pojo generation
   * needs to specify inputs and outputs
* docker compose build
   * could use the bmuschko docker plugin
* integration tests that compile incrementally but don't run incrementally
* node
   * needs to specify inputs and outputs

tl;dr: various assumptions about our project structure that are baked in due to the desire to have self-contained connector builds cause a nightmare for the core platform.

**Comment by sherifnada** [comment]
also: 

* find a way to make python build incrementally. linkedin plugin is an option but probably requires a single venv.
* find a way to make docker images build incrementally. there’s another plugin which might work (which one? )
* we want an easy way to see the different build pipelines and their status/health
* we want an easy way to run on forks

**Comment by davinchia** [comment]
@benmoriceau I know you did some good work. what if left here?
